### DISCUSSION TITLE
WITH RECURSIVE not working for me

### DISCUSSION THREAD

**Discussion by JerContact** [simolus3/drift#2448] [Q&A]
Hello! I looking in your code and it seems that recursive WITH is in there, but the .g file won't generate this code:

 ```SQL
CREATE TRIGGER CHECK_STUFF
AFTER INSERT ON Main
FOR EACH ROW
BEGIN
    WITH cte AS (
        SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = NEW.FolderTreeId) AS myID, NEW.FolderTreeId AS id
        UNION ALL
        SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = myID), myID
        FROM cte
        WHERE myID IS NOT NULL
    )
    UPDATE FOLDERTREE SET RefCount = RefCount + 1 WHERE id IN (SELECT id FROM cte);
END;
```

If i go into DBBrowser it's working as expected, but for some reason drift won't auto generate this

**Comment by simolus3** [comment]
> but for some reason drift won't auto generate this

Is there an error being emitted in `build_runner build` at all, or is the trigger just silently missing? Can you share a drift file (with the relevant tables) that reproduces this behavior?

**Comment by JerContact** [comment]
I do get a warning sometimes

[WARNING] drift_dev on lib/data/dataproviders/drift/drift_audio_metadata_database.drift:
Could not analyze DriftElementId(package:slt_flutter/data/dataproviders/drift/drift_audio_metadata_database.drift, CHECK_STUFF)     
Null check operator used on a null value

**Reply by simolus3** [reply]
In that case, could you re-run the build with `build_runner build --verbose` and post the stack trace attached to that message?

**Comment by JerContact** [comment]
Here you are!

```
[WARNING] drift_dev on lib/data/dataproviders/drift/drift_audio_metadata_database.drift:
Could not analyze DriftElementId(package:slt_flutter/data/dataproviders/drift/drift_audio_metadata_database.drift, CHECK_STUFF)     
Null check operator used on a null value
package:sqlparser/src/analysis/steps/column_resolver.dart 218:54            ColumnResolver._handle.<fn>
package:sqlparser/src/ast/common/queryables.dart 14:21                      Queryable.when
package:sqlparser/src/analysis/steps/column_resolver.dart 212:15            ColumnResolver._handle
package:sqlparser/src/analysis/steps/column_resolver.dart 271:7             ColumnResolver._resolveSelect
package:sqlparser/src/analysis/steps/column_resolver.dart 16:5              ColumnResolver.visitSelectStatement
package:sqlparser/src/ast/statements/select.dart 54:20                      SelectStatement.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/ast/visitor.dart 572:12                               RecursiveVisitor.defaultNode
package:sqlparser/src/ast/visitor.dart 187:12                               RecursiveVisitor.visitCompoundSelectPart
package:sqlparser/src/ast/statements/select.dart 250:20                     CompoundSelectPart.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/analysis/steps/column_resolver.dart 24:5              ColumnResolver.visitCompoundSelectStatement
package:sqlparser/src/ast/statements/select.dart 105:20                     CompoundSelectStatement.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/analysis/steps/column_resolver.dart 41:5              ColumnResolver.visitCommonTableExpression
package:sqlparser/src/ast/clauses/with.dart 63:20                           CommonTableExpression.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/ast/visitor.dart 572:12                               RecursiveVisitor.defaultNode
package:sqlparser/src/ast/visitor.dart 213:12                               RecursiveVisitor.visitWithClause
package:sqlparser/src/ast/clauses/with.dart 15:20                           WithClause.accept
package:sqlparser/src/analysis/steps/column_resolver.dart 82:19             ColumnResolver.visitUpdateStatement
package:sqlparser/src/ast/statements/update.dart 53:20                      UpdateStatement.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/ast/visitor.dart 572:12                               RecursiveVisitor.defaultNode
package:sqlparser/src/ast/visitor.dart 382:12                               RecursiveVisitor.visitBlock
package:sqlparser/src/ast/statements/block.dart 16:20                       Block.accept
package:sqlparser/src/ast/visitor.dart 598:12                               VisitChildrenExtension.visitList
package:sqlparser/src/ast/visitor.dart 592:12                               VisitChildrenExtension.visitChildren
package:sqlparser/src/analysis/steps/column_resolver.dart 193:5             ColumnResolver.visitCreateTriggerStatement
package:sqlparser/src/ast/statements/create_trigger.dart 32:20              CreateTriggerStatement.accept
package:sqlparser/utils/find_referenced_tables.dart 165:11                  FindReferenceAnalysis.findReferencedSchemaTables        
package:drift_dev/src/analysis/resolver/drift/element_resolver.dart 171:40  DriftElementResolver.resolveTableReferences
package:drift_dev/src/analysis/resolver/drift/trigger.dart 19:30            DriftTriggerResolver.resolve
package:drift_dev/src/analysis/resolver/resolver.dart 68:37                 DriftResolver.resolveDiscovered
package:drift_dev/src/analysis/driver/driver.dart 206:26                    DriftAnalysisDriver._analyzePrepared
package:drift_dev/src/analysis/driver/driver.dart 237:5                     DriftAnalysisDriver.resolveElements
package:drift_dev/src/backends/build/analyzer.dart 36:21                    DriftAnalyzer.build
```

**Reply by simolus3** [reply]
Thanks! I can reproduce this problem and I think I've fixed it in 9154f60dfd6d9aa977573b9b5b9efa579370c837. Could you add this to your pubspec and let me know if it fixes the problem for you / whether there are any regressions?

```yaml
dependency_overrides:
  sqlparser:
    git:
      url: https://github.com/simolus3/drift.git
      ref: develop
      path: sqlparser

```

**Reply by JerContact** [reply]
I seem to get a sql syntax error no matter what I put, and it seems to work fine on db browser execute, is there an error I'm just not seeing?

Exception has occurred.
DriftRemoteException (SqliteException(1): while executing, near "UPDATE": syntax error, SQL logic error (code 1)
  Causing statement: CREATE TRIGGER CHECK_STUFF AFTER INSERT ON AudioMetadataMain BEGIN WITH RECURSIVE cte AS (SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = NEW.FolderTreeId) AS myID UNION ALL SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = myID) FROM cte WHERE myID IS NOT NULL) UPDATE FOLDERTREE SET RefCount = 1;END, parameters: )


Here is the stack for the error:
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: SqliteException(1): while executing, near "UPDATE": syntax error, SQL logic error (code 1)
  Causing statement: CREATE TRIGGER CHECK_STUFF AFTER INSERT ON AudioMetadataMain BEGIN WITH RECURSIVE cte AS (SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = NEW.FolderTreeId) AS myID UNION ALL SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = myID) FROM cte WHERE myID IS NOT NULL) UPDATE FOLDERTREE SET RefCount = 1;END, parameters:
#0      DriftCommunication.request
#1      _RemoteQueryExecutor.ensureOpen
#2      LazyDatabase.ensureOpen.<anonymous closure>
<asynchronous suspension>
#3      DriftAudioMetadataDatabase.searchMetadata
<asynchronous suspension>
#4      DriftDatabaseRepository.searchMetadataEmptyText
<asynchronous suspension>
#5      DriftDatabaseBloc._textSearch
<asynchronous suspension>
#6      DriftDatabaseBloc._onTextSearched
<asynchronous suspension>
#7      Bloc.on.<anonymous closure>.handleEvent
<asynchronous suspension>

  

Here is just a test to see if the trigger will actually do anything, but i get the error above 

CREATE TRIGGER CHECK_STUFF
AFTER INSERT ON AudioMetadataMain
FOR EACH ROW
BEGIN
    WITH RECURSIVE cte AS (
        SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = NEW.FolderTreeId) AS myID
        UNION ALL
        SELECT (SELECT ParentId FROM FOLDERTREE WHERE id = myID)
        FROM cte
        WHERE myID IS NOT NULL
    )
    UPDATE FOLDERTREE SET RefCount = 1;
END;

**Reply by simolus3** [reply]
Well this is disappointing, but the sqlite3 docs mention that [WITH clauses cannot be used in triggers](https://sqlite.org/lang_with.html#limitations_and_caveats). So that explains the syntax error, I'll add this error to drift as well.

**Reply by JerContact** [reply]
ahhhhh, that makes sense!  Thanks for looking into that!!
